git-info: temporarily set IFS to $'\t' for tamper-proof extraction of ahead and behind

zmwangx opened the pull request:
The problem with the current implementation of `ahead` and `behind` extraction in `git-info` is that it is not tamper-proof against a modified `IFS`. The output format of `git rev-list --count --left-right ...` is `ahead<TAB>behind`, and `ahead` and `behind` are parsed from that output through word splitting; but if `IFS` doesn't contain the TAB character, then splitting will fail, leading to angry bad math expression errors.

To fix this problem, I'm temporarily setting `IFS` to `$'\t'`, then restore it when done.

Another thought is to use parameter expansion flags `(ws:<separator>:)`. The code would be cleaner, but I tried to use `$'\t'` as separator with no luck (unless I hard code the tab, which I'm afraid is not so readable and would easily fall victim to editor commands like `untabify`). It seems that quoting doesn't work inside the flags, for some reason. Thoughts are welcome.

Here's a small test case:

``` zsh
> string=$'1\t2'
> echo $string[(ws: :)1]  # between the two colons is a literal tab
1
> echo $string[(ws:$'\t':)1]
1   2
```

sorin-ionescu commented:
While your first example may fall victim to the tabs to spaces command, it will be caught by `diff`; so, it's not completely hidden.

zmwangx commented:
Yes it's not, so you're actually in favor of that? The problem I had in mind was that someone reading the source code (say, on the GitHub web interface) might have a hard time realizing that it's a tab rather than whatever number of spaces.

sorin-ionescu commented:
A comment that specifies that it's a tab ought to fix that. It would not be the first time we had [unreadable character in Prezto](https://github.com/sorin-ionescu/prezto/blob/master/modules/editor/init.zsh#L66).

zmwangx commented:
Sure, rebased.

zmwangx commented:
Switched to `(pws:\t:)` to avoid the literal tab character. Previously I wasn't aware of the parameter expansion [flag `p`](http://zsh.sourceforge.net/Doc/Release/Expansion.html#Parameter-Expansion-Flags):

> Recognize the same escape sequences as the `print` builtin in string arguments to any of the flags described below that follow this argument.

sorin-ionescu commented:
Nice. Is there anywhere else in Prezto where literal characters are used?

zmwangx commented:
I grepped for all characters outside the ASCII printable range (with GNU grep):

```
> grep -rPn --exclude-dir=.git --exclude=.gitmodules "[^[:print:]]" .
./modules/editor/init.zsh:66:    key_info[$key]='�'
./modules/prompt/functions/prompt_cloud_setup:26:#   If these options are not provided, the symbol defaults to "☁" with colors
./modules/prompt/functions/prompt_cloud_setup:59:If these options are not provided, the symbol defaults to ☁ with colors cyan
./modules/prompt/functions/prompt_cloud_setup:70:    prompt_preview_theme 'cloud' "✯"
./modules/prompt/functions/prompt_cloud_setup:85:    prefix='☁'
./modules/prompt/functions/prompt_cloud_setup:109:  zstyle ':prezto:module:git:info:dirty' format "%%B%F{$secondary_color}]%f%%b %F{yellow}⚡%f"
./modules/prompt/functions/prompt_damoekri_setup:61:  zstyle ':prezto:module:git:info:clean' format ' %F{green}✔%f'
./modules/prompt/functions/prompt_damoekri_setup:62:  zstyle ':prezto:module:git:info:dirty' format ' %F{red}✗%f'
./modules/prompt/functions/prompt_kylewest_setup:45:  zstyle ':prezto:module:editor:info:keymap:primary' format "%B%F{green}❯%f%b"
./modules/prompt/functions/prompt_kylewest_setup:46:  zstyle ':prezto:module:editor:info:keymap:alternate' format "%B%F{magenta}❮%f%b"
./modules/prompt/functions/prompt_minimal_setup:15:    hook_com[unstaged]='%F{red}●%f'
./modules/prompt/functions/prompt_minimal_setup:38:  zstyle ':vcs_info:*' stagedstr '%F{green}●%f'
./modules/prompt/functions/prompt_minimal_setup:39:  zstyle ':vcs_info:*' unstagedstr '%F{yellow}●%f'
./modules/prompt/functions/prompt_nicoulaj_setup:40:  local user_char='❯'
./modules/prompt/functions/prompt_nicoulaj_setup:41:  local root_char='❯❯❯'
./modules/prompt/functions/prompt_paradox_setup:21:_prompt_paradox_segment_separator=''
./modules/prompt/functions/prompt_paradox_setup:48:  prompt_paradox_start_segment black default '%(?::%F{red}✘ )%(!:%F{yellow}⚡ :)%(1j:%F{cyan}⚙ :)%F{blue}%n%F{red}@%F{green}%m%f'
./modules/prompt/functions/prompt_paradox_setup:122:  zstyle ':prezto:module:editor:info:keymap:primary' format '%B%F{blue}❯%f%b'
./modules/prompt/functions/prompt_paradox_setup:123:  zstyle ':prezto:module:editor:info:keymap:primary:overwrite' format '%F{red}♺%f'
./modules/prompt/functions/prompt_paradox_setup:124:  zstyle ':prezto:module:editor:info:keymap:alternate' format '%B%F{red}❮%f%b'
./modules/prompt/functions/prompt_paradox_setup:128:  zstyle ':prezto:module:git:info:action' format ' ⁝ %s'
./modules/prompt/functions/prompt_paradox_setup:129:  zstyle ':prezto:module:git:info:added' format ' ✚'
./modules/prompt/functions/prompt_paradox_setup:130:  zstyle ':prezto:module:git:info:ahead' format ' ⬆'
./modules/prompt/functions/prompt_paradox_setup:131:  zstyle ':prezto:module:git:info:behind' format ' ⬇'
./modules/prompt/functions/prompt_paradox_setup:132:  zstyle ':prezto:module:git:info:branch' format ' %b'
./modules/prompt/functions/prompt_paradox_setup:133:  zstyle ':prezto:module:git:info:commit' format '➦ %.7c'
./modules/prompt/functions/prompt_paradox_setup:134:  zstyle ':prezto:module:git:info:deleted' format ' ✖'
./modules/prompt/functions/prompt_paradox_setup:135:  zstyle ':prezto:module:git:info:dirty' format ' ⁝'
./modules/prompt/functions/prompt_paradox_setup:136:  zstyle ':prezto:module:git:info:modified' format ' ✱'
./modules/prompt/functions/prompt_paradox_setup:138:  zstyle ':prezto:module:git:info:renamed' format ' ➙'
./modules/prompt/functions/prompt_paradox_setup:140:  zstyle ':prezto:module:git:info:unmerged' format ' ═'
./modules/prompt/functions/prompt_peepcode_setup:45:  zstyle ':prezto:module:git:info:dirty' format ' %F{8}✗%f'
./modules/prompt/functions/prompt_peepcode_setup:54:%(?.%F{green}${1:-☻ }%f.%F{red}${1:-☻ }%f) "
./modules/prompt/functions/prompt_peepcode_setup:65:If this option is not provided, the symbol defaults to ☻.
./modules/prompt/functions/prompt_peepcode_setup:79:    prompt_preview_theme 'peepcode' "❯"
./modules/prompt/functions/prompt_skwp_setup:61:  zstyle ':prezto:module:git:info:added' format "${_prompt_skwp_colors[5]}●%f"
./modules/prompt/functions/prompt_skwp_setup:62:  zstyle ':prezto:module:git:info:deleted' format "${_prompt_skwp_colors[2]}●%f"
./modules/prompt/functions/prompt_skwp_setup:63:  zstyle ':prezto:module:git:info:modified' format "${_prompt_skwp_colors[4]}●%f"
./modules/prompt/functions/prompt_skwp_setup:64:  zstyle ':prezto:module:git:info:untracked' format "${_prompt_skwp_colors[3]}●%f"
./modules/prompt/functions/prompt_smiley_setup:60:  PROMPT='$python_info[virtualenv]$ruby_info[version]${git_info:+${(e)git_info[prompt]}} %B%c%b %(?:%F{green}ツ%f:%F{red}✖%f) '
./modules/prompt/functions/prompt_sorin_setup:83:  RPROMPT='${editor_info[overwrite]}%(?:: %F{1}⏎%f)${VIM:+" %B%F{6}V%f%b"}'
./modules/prompt/functions/prompt_sorin_setup:111:  zstyle ':prezto:module:editor:info:keymap:primary' format ' %B%F{1}❯%F{3}❯%F{2}❯%f%b'
./modules/prompt/functions/prompt_sorin_setup:112:  zstyle ':prezto:module:editor:info:keymap:primary:overwrite' format ' %F{3}♺%f'
./modules/prompt/functions/prompt_sorin_setup:113:  zstyle ':prezto:module:editor:info:keymap:alternate' format ' %B%F{2}❮%F{3}❮%F{1}❮%f%b'
./modules/prompt/functions/prompt_sorin_setup:118:  zstyle ':prezto:module:git:info:added' format ' %%B%F{2}✚%f%%b'
./modules/prompt/functions/prompt_sorin_setup:119:  zstyle ':prezto:module:git:info:ahead' format ' %%B%F{13}⬆%f%%b'
./modules/prompt/functions/prompt_sorin_setup:120:  zstyle ':prezto:module:git:info:behind' format ' %%B%F{13}⬇%f%%b'
./modules/prompt/functions/prompt_sorin_setup:123:  zstyle ':prezto:module:git:info:deleted' format ' %%B%F{1}✖%f%%b'
./modules/prompt/functions/prompt_sorin_setup:124:  zstyle ':prezto:module:git:info:modified' format ' %%B%F{4}✱%f%%b'
./modules/prompt/functions/prompt_sorin_setup:126:  zstyle ':prezto:module:git:info:renamed' format ' %%B%F{5}➜%f%%b'
./modules/prompt/functions/prompt_sorin_setup:127:  zstyle ':prezto:module:git:info:stashed' format ' %%B%F{6}✭%f%%b'
./modules/prompt/functions/prompt_sorin_setup:128:  zstyle ':prezto:module:git:info:unmerged' format ' %%B%F{3}═%f%%b'
./modules/prompt/functions/prompt_sorin_setup:129:  zstyle ':prezto:module:git:info:untracked' format ' %%B%F{7}◼%f%%b'
./modules/prompt/functions/prompt_steeef_setup:18:    branch_format="(${_prompt_steeef_colors[1]}%b%f%u%c${_prompt_steeef_colors[4]}●%f)"
./modules/prompt/functions/prompt_steeef_setup:72:  local unstaged_format="${_prompt_steeef_colors[2]}●%f"
./modules/prompt/functions/prompt_steeef_setup:73:  local staged_format="${_prompt_steeef_colors[5]}●%f"
./README.md:1:Prezto — Instantly Awesome Zsh
```

Seems all legit use, can't be replaced.

Is there any ASCII printable character other than the tab that's confusing as a "literal"? I can't think of any. Just grepping for tabs returns nothing (I only grepped for tabs between two non-whitespace characters; otherwise we'd get a whole lot of tabs from indentation):

```
> grep -rPn --exclude-dir=.git --exclude=.gitmodules "(?<=\S)\t(?=\S)" . || echo "No offenders."
No offenders.
```

belak commented:
Strangely, I just noticed this because it was closed... it seems useful so I'm going to re-open and merge it if you don't have any objections.

zmwangx commented:
I was just closing my ancient PRs accords the site. If it’s still relevant and someon’s willing to take care of it, there’s no reason to object...

belak commented:
Thanks again for submitting this - sorry it took us so long to get to it.
